cluster-api/machines-api/types.go

```diff
@@ -0,0 +1,160 @@
+package machines
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/api"
+)
+
+// Machine represents a single Node that should exist (whether it does or
+// not yet). In this model, there is no grouping of nodes to scale with a
+// numeric field. Each Machine exists independently, and grouping can only
+// be inferred via label selectors.
+//
+// In order for a new Node to be created, one can generically create a new
+// Machine object, possibly copying the spec from an existing Machine
+// or a template. To scale down the cluster, delete specific instances of
+// Machines and the underlying Nodes will be unregistered/deprovisioned.
+// Separate provider-specific controllers will watch Machine objects they can
+// act on (like a GCE cloud controller watching for only Machines destined for
+// GCE) and take the appropriate actions.
+//
+// Any updates to the MachineSpec will be actuated to change the Node in
+// place or replace the Node with one conforming to the spec. In this model,
+// the fact that a controller is able to upgrade a node via in-place upgrades
+// or via a cloud replacement is an implementation detail without API controls.
+//
+// It is recommended, but not required, that provider-specific controllers add
+// finalizers to Machine objects so that they can be triggered on deletion to
+// release the necessary external resources, reporting any errors encountered.
+type Machine struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec   MachineSpec
+	Status MachineStatus
+}
+
+type MachineSpec struct {
+	// This ObjectMeta will autopopulate the Node created. Use this to
+	// indicate what labels, annotations, name prefix, etc., should be used
+	// when creating the Node.
+	// +optional
+	metav1.ObjectMeta
+
+	// Provider-specific serialized configuration to use during node
+	// creation. It is recommended that providers maintain their own
+	// versioned API types that should be serialized/deserialized from this
+	// field, akin to component config.
+	// +optional
+	ProviderConfig string
+
+	// A list consisting of "Master" and/or "Node".
+	//
+	//                 +-----------------------+------------------------+
+	//                 | Master present        | Master absent          |
+	// +---------------+-----------------------+------------------------|
+	// | Node present: | Install control plane | Join the cluster as    |
+	// |               | and be schedulable    | just a node            |
```

Comment:
I'm not sure schedulable exists as a concept any more... AIUI the masters are supposed to be schedulable, but tainted so user pods aren't scheduled to it.  But e.g. a monitoring daemonset or networking daemonset should tolerate the taint and thus be on the master.

Reply:
Ah, this is poor wording on my part. If I replace "schedulable" with "untainted" and "unschedulable' with "tainted," is that sufficient? Or should we not differentiate between (`Master`) and (`Master`, `Node`)? Or do you think there's a better way to represent the desire to install the control plane altogether?

Reply:
From what @justinsb said, it sounds like there shouldn't be a distinction between Master and Node, and I think I agree with this position. Is there any scenario where an un-tainted master would be significantly preferable to a tainted one?